File: 20181202/checksum.js
"use strict";
let fs = require('fs')

function getFileContents(filename) {
  try {
    return fs.readFileSync(filename, 'utf8')
  } catch (e) {
    console.log(e);
  };
}

function createMap(s) {
  let m = new Map();

  for (let i = 0; i < s.length; i++) {
    let c = s[i];
    if (m.has(c)) {
      m.set(c, m.get(c) + 1);
    } else {
      m.set(c, 1);
    }
  }

  return m;
}

function evalCharCount(map, count) {
  for (let [k, v] of map) {
    if (v === count) {
      return true;
    }
  }
  return false;
}

let inputMaps = getFileContents("input.txt").split("\n").map(createMap);

let doublesCount = 0, triplesCount = 0
inputMaps.forEach(m => {
  if (evalCharCount(m, 2)) {
    doublesCount++;
  }

  if (evalCharCount(m, 3)) {
    triplesCount++;
  }
})

console.log(`doubles ${doublesCount} || triples ${triplesCount}`);
console.log(`checksum = ${doublesCount * triplesCount}`);

File: 20181202/checksum.go
package main

import (
	"fmt"
	"strings"
)

func createMap(str string) map[string]int {
	// create map of character + count of character within string
	// ie ccabcacc -->
	// {
	//   a: 2,
	//   b: 1,
	//   c: 5,
	// }

	m := make(map[string]int)

	for _, c := range str {
		m[string(c)]++
	}

	return m
}

func evalCharCount(m map[string]int, count int) bool {
	for _, v := range m {
		if v == count {
			return true
		}
	}
	return false
}

func Checksum() {
	fmt.Println("-----------------------------")
	inputs := strings.Split(GetFileAsString("input.txt"), "\n")

	var doublesCount, triplesCount int = 0, 0
	for i := 0; i < len(inputs); i++ {
		m := createMap(inputs[i])

		if evalCharCount(m, 2) {
			doublesCount++
		}

		if evalCharCount(m, 3) {
			triplesCount++
		}
	}

	fmt.Printf("doubles: %d triples %d\n", doublesCount, triplesCount)
	fmt.Println("checksum is ", doublesCount*triplesCount)
	fmt.Println("-----------------------------")
}
